password_enhancements.module

```diff
@@ -0,0 +1,471 @@
+ *
+ * This callback runs only if the password field was set to always require and
+ * the verify_email set to true.
+ *
+ * @param array $form
+ *   The form.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The current form state.
+ */
+function password_enhancements_form_user_register_form_alter_password_required_submit(array &$form, FormStateInterface $form_state): void {
+  /** @var \Drupal\user\UserInterface $user */
+  if (!empty($form['account']['pass'])) {
+    /** @var \Drupal\Core\Entity\EntityFormInterface $form_object */
+    $form_object = $form_state->getFormObject();
+    $user = $form_object->getEntity();
+
+    try {
+      $user->setPassword($form_state->getUserInput()['pass']['pass1'])->save();
+    }
+    catch (Exception $e) {
+      Drupal::service('logger.password_enhancements')->exception($e);
+      Drupal::messenger()->addError(t('Failed to set your password, please <a href="@url">reset your password</a>.', [
+        '@url' => Url::fromRoute('user.pass')->toString(),
+      ]));
+    }
+  }
+}
+
+/**
+ * Ajax callback for updating the constraints on role change.
+ *
+ * @param array $form
+ *   The form.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The current form state.
+ *
+ * @return array
+ *   Renderable form array for password policy constraints.
+ */
+function password_enhancements_update_constraints_by_roles(array $form, FormStateInterface $form_state): array {
+  return $form['account']['password_policy_constraints'];
+}
+
+/**
+ * Implements hook_entity_base_field_info().
+ */
+function password_enhancements_entity_base_field_info(EntityTypeInterface $entity_type): array {
+  $fields = [];
+
+  if ($entity_type->id() === 'user') {
+    $fields['password_enhancements_password_changed_date'] = BaseFieldDefinition::create('datetime')
+      ->setLabel(t('Last password change'))
+      ->setDescription(t('The last time when the user has changed the password.'))
+      ->setDefaultValue(0);
+
+    $fields['password_enhancements_password_change_required'] = BaseFieldDefinition::create('boolean')
+      ->setLabel(t('Password change required'))
+      ->setDescription(t('Forces the user to change the password.<br><strong>Warning:</strong> If it is checked, the user will be redirected to the password change form.'))
+      ->setDefaultValue(0)
+      ->setDisplayOptions('form', [
+        'type' => 'boolean_checkbox',
+        'settings' => [
+          'display_label' => TRUE,
+        ],
+        'weight' => 99,
+      ]);
+  }
+
+  return $fields;
+}
+
+/**
+ * Implements hook_ENTITY_TYPE_presave().
+ */
+function password_enhancements_user_presave(UserInterface $user): void {
+  // Update the last password changed date if the password was changed.
+  if (empty($user->original) || $user->original->pass->value !== $user->pass->value) {
+    $user->password_enhancements_password_changed_date->setValue(Drupal::time()->getRequestTime());
```

Comment:
Nit: `    $user->get('password_enhancements_password_changed_date')->setValue(Drupal::time()->getRequestTime());` because the `get()` method is defined on the `UserInterface` but the `password_enhancements_password_changed_date` does not. (Bad Drupal DX)